### PR TITLE
Team Member Submission -> PHP

### PR DESCRIPTION
**Pull request by crast**
* Get rid of SQL based submission
* Write some functional tests

### CODE REVIEW

File: src/app/Domain/TeamMember.php [hunk 1]
```diff
@@ -253,3 +253,6 @@ public function getAssociatedPerson()
     }
 
 }
+{
```
Comment (pdarg) [inline src/app/Domain/TeamMember.php]: what's this little guy?

File: src/app/Api/SubmissionCore.php [hunk 2]
```diff
@@ -684,6 +590,69 @@ public function calculateProgramLeaderAttending(Models\Center $center, Carbon $r
...
+    }
+
+    public function submitTeamMembers(Models\Center $center, Carbon $reportingDate, Models\StatsReport $statsReport, $teamMembers)
+    {
+        $debug_message = '';
+        foreach ($teamMembers as $tmDomain) {
+            list($tm, $person) = $this->ensureTeamMember($center, $tmDomain);
+            if ($tmDomain->id < 0) {
+                // If ID < 0, then we need to update any existing stash with the newly assigned team member ID
+                Models\SubmissionData::centerDate($center, $reportingDate)->typeId('team_member', $tmDomain->id)->update(['id' => $tm->id]);
```
Comment (pdarg) [inline src/app/Api/SubmissionCore.php]: You also need to update `stored_id`

File: src/app/Api/SubmissionCore.php [hunk 3]
```diff
@@ -684,6 +590,69 @@ public function calculateProgramLeaderAttending(Models\Center $center, Carbon $r
...
+
+    /**
+     * Save a model if it's dirty.
+     *
+     * This prevents updating timestamps when there are no changes to make.
+     * @param  Model $model The model.
+     */
+    protected static function saveIfDirty(Model $model)
+    {
+        if ($model->isDirty()) {
```
Comment (pdarg) [inline src/app/Api/SubmissionCore.php]: This check is already done while saving:
https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Eloquent/Model.php#L1539-L1541

Comment (crast) [reply src/app/Api/SubmissionCore.php]: hmm, I must'be been looking at the wrong code then... it seemed to me like it updated timestamps and then did the dirty check, and timestamps would therefore be dirty

File: src/app/Api/SubmissionCore.php [hunk 4]
```diff
@@ -711,6 +608,78 @@ public function calculateProgramLeaderAttending(Models\Center $center, Carbon $r
...
+            ]);
+            $tm->setRelation('person', $person); // To avoid an extra lookup
+
+            return [$tm, $person];
+        } else {
+            $tm = Models\TeamMember::findOrFail($tmDomain->id);
+
+            // Check for team member person overrides
+            if ($tmDomain->_personId != null && $tm->personId != $tmDomain->_personId) {
+                $tm->personId = $person->id;
```
Comment (pdarg) [inline src/app/Api/SubmissionCore.php]: do we want to support this? I think _personId is useful when creating a new object, but I'm not sure we'll ever want to change the person associated with an existing TeamMember object

File: src/app/Api/SubmissionCore.php (hunk 3 above)
Comment (pdarg) [reply src/app/Api/SubmissionCore.php]: in that code, it's checking for dirty first, then running `$this->updateTimestamps()` if it's dirty

File: src/app/Api/SubmissionCore.php [hunk 5]
```diff
@@ -711,6 +608,78 @@ public function calculateProgramLeaderAttending(Models\Center $center, Carbon $r
...
+    }
+
+    public function submitTeamMembers(Models\Center $center, Carbon $reportingDate, Models\StatsReport $statsReport, $teamMembers)
+    {
+        $debug_message = '';
+        foreach ($teamMembers as $tmDomain) {
+            list($tm, $person) = $this->ensureTeamMember($center, $tmDomain);
+            if ($tmDomain->id < 0) {
+                // If ID < 0, then we need to update any existing stash with the newly assigned team member ID
+                Models\SubmissionData::centerDate($center, $reportingDate)->typeId('team_member', $tmDomain->id)->update(['stored_id' => $tm->id]);
```
Comment (pdarg) [inline src/app/Api/SubmissionCore.php]: I think you also need to update the id inside the stashes data field. Theres an existing method that does both in there somewhere

**Pull request merged**